core/src/worker/workflow/mod.rs

```diff
@@ -317,6 +318,180 @@ impl Workflows {
+            } => {
+                let reserved_act_permits =
+                    self.reserve_activity_slots_for_outgoing_commands(commands.as_mut_slice());
+                debug!(commands=%commands.display(), query_responses=%query_responses.display(),
+                           messages=%messages.display(), force_new_wft,
+                           "Sending responses to server");
+                if let Some(default_vb) = self.default_versioning_behavior.as_ref() {
+                    if versioning_behavior == VersioningBehavior::Unspecified {
+                        versioning_behavior = *default_vb;
+                    }
+                }
+                let mut completion = WorkflowTaskCompletion {
+                    task_token: task_token.clone(),
+                    commands,
+                    messages,
+                    query_responses,
+                    sticky_attributes: None,
+                    return_new_workflow_task: true,
+                    force_create_new_workflow_task: force_new_wft,
+                    sdk_metadata,
+                    metering_metadata: MeteringMetadata {
+                        nonfirst_local_activity_execution_attempts: self
+                            .local_act_mgr
+                            .get_nonfirst_attempt_count(run_id)
+                            as u32,
+                    },
+                    versioning_behavior,
+                };
+                let sticky_attrs = self.sticky_attrs.clone();
+                // Do not return new WFT if we would not cache, because returned new WFTs are
+                // always partial.
+                if sticky_attrs.is_none() {
+                    completion.return_new_workflow_task = false;
+                }
+                completion.sticky_attributes = sticky_attrs;
+
+                let mut reset_last_started_to = None;
+                self.handle_wft_reporting_errs(run_id, || async {
+                    match self.client.complete_workflow_task(completion).await {
+                        Ok(response) => {
+                            if response.reset_history_event_id > 0 {
+                                reset_last_started_to = Some(response.reset_history_event_id);
+                            }
+                            if let Some(wft) = response.workflow_task {
+                                *wft_from_complete = Some(validate_wft(wft)?);
+                            }
+                            self.handle_eager_activities(
+                                reserved_act_permits,
+                                response.activity_tasks,
+                            );
+                        }
+                        Err(e) if e.metadata().contains_key("message-too-large") => {
+                            let failure = Failure {
+                                failure: Some(
+                                    temporal_sdk_core_protos::temporal::api::failure::v1::Failure {
```

Comment:
Arguably we should put an application error info in here with an error type string of `GrpcMessageTooLarge` or something. When we build rules to react to task failure (e.g. ones that pause or notify), users will need something to uniquely identify the error. The task failed cause enum will do this too, but I suspect most users will inspect the failure. Not required of course.